src/charm.py

```diff
@@ -872,16 +874,40 @@ def enable_disable_extensions(self, database: str = None) -> None:
                 for ext in spi_module:
                     extensions[ext] = enable
                 continue
-            if extension in plugins_exception:
-                extension = plugins_exception[extension]
+            extension = plugins_exception.get(extension, extension)
+            if self._check_extension_dependencies(extension, enable):
+                self.unit.status = BlockedStatus(EXTENSIONS_BLOCKING_MESSAGE)
+                return
             extensions[extension] = enable
+        if self.is_blocked and self.unit.status.message == EXTENSIONS_BLOCKING_MESSAGE:
+            self.unit.status = ActiveStatus()
         self.unit.status = WaitingStatus("Updating extensions")
         try:
             self.postgresql.enable_disable_extensions(extensions, database)
         except PostgreSQLEnableDisableExtensionError as e:
             logger.exception("failed to change plugins: %s", str(e))
         self.unit.status = original_status
 
+    def _check_extension_dependencies(self, extension: str, enable: bool) -> bool:
+        skip = False
+        if enable and extension in REQUIRED_PLUGINS:
+            for ext in REQUIRED_PLUGINS[extension]:
+                if not self.config[f"plugin_{ext}_enable"]:
+                    skip = True
+                    logger.exception(
+                        "cannot enable %s, extension required %s to be enabled before",
+                        extension,
+                        ext,
+                    )
+        elif not enable and extension in DEPENDENCY_PLUGINS:
+            for n, v in REQUIRED_PLUGINS.items():
+                if extension in v and self.config["plugin_" + n + "_enable"]:
+                    skip = True
+                    logger.exception(
+                        f"{n} require plugin {extension}. Switch off first {n} before disable {extension}"
+                    )
```

Comment:
I don't think we need this loop. The first one will run over all the plugins and verify them anyway.